### ISSUE TITLE
configure default inheritance type

### ISSUE THREAD

**Issue by wumpz** [patrodyne/hisrc-hyperjaxb#1]
**Thx for keeping hyperjaxb alive for more Java and Jakarta versions.**

This **inheritance** configuration is an open problem from the original project.

Inheritance is at the moment defaulted to this annotation: `@Inheritance(strategy = InheritanceType.JOINED)`. 

Here are my problems: 
1. First of all I want to change this InheritanceType.
2. HyperJaxb does put this annotation to each **base** class, even if there are no inherited classes. The JPS specification says for this JOINED type something like: 

> If we didn't pass the name of discriminator column and its value then JPA consider it by default. It considers **DTYPE** as discriminator column name and the name of the entity as discriminator value.

**Hibernate** fortunately skips this but **EclipseLink** produces for each of those classes a *DTYPE* column of strangely **varchar(31)**, which is spec as well. 

This is quite annoying. :(

Since I have a schema that produces over 1500 classes this should not be a manual task. ;)

**Comment by patrodyne** [comment]
This [solution](https://github.com/patrodyne/hisrc-hyperjaxb/tree/master/ejb/assembly/samples/publication#solution) and [sample project (zip)][20] demonstrates the use of bindings to *change the inheritance type*:

[Publication.xjb][39]
~~~
<hj:default-entity>
    <orm:inheritance strategy="SINGLE_TABLE"/>
</hj:default-entity>
~~~

Also, the demo includes an [approach][99] to remove the `DTYPE` column from non-inherited entities.




[20]: https://github.com/patrodyne/hisrc-hyperjaxb/releases/download/2.1.0/hisrc-hyperjaxb-ejb-sample-publication-2.1.0-mvn-src.zip
[39]: https://github.com/patrodyne/hisrc-hyperjaxb/tree/master/ejb/assembly/samples/publication/src/main/resources/Publication.xjb
[99]: https://github.com/patrodyne/hisrc-hyperjaxb/tree/master/ejb/assembly/samples/publication#dtype-omit-inheritance-from-childless-roots

**Comment by wumpz** [comment]
Thx again for keeping this alive. 
And thx for the solution. I will try it.

**Comment by patrodyne** [comment]
**Note:** Looks like Hibernate is aligning towards the JPA/EclipseLink behavior: [HHH-6911](https://hibernate.atlassian.net/browse/HHH-6911).

~~~
Legacy Hibernate behavior was to ignore the @DiscriminatorColumn.
However, as part of issue HHH-6911 we now apply the explicit
@DiscriminatorColumn. If you would prefer the legacy behavior,
enable the `hibernate.discriminator.ignore_explicit_for_joined`
setting:

    hibernate.discriminator.ignore_explicit_for_joined=true
~~~

An open question is "Can/should HyperJAXB attempt to identify non-inherited (i.e. non-base) entities and omit the `@Inheritance` annotation in order to reduce `DTYPE` columns in the generated tables?" One consideration is the use of sub-entities outside of the XML schema and/or future sub-entities that may require the `@Inheritance` to be there. I'll experiment.

**Comment by patrodyne** [comment]
In **HyperJAXB** [Release 2.1.1][1], the commit [1a269e2][94] modified `EntityMapping.java` to check when a root class is not inherited by some other class in the JAXB context *and* to determine if the root class contains at least one basic field. When these conditions are satisfied, the `@Inheritance` annotation is omitted from the root class.

New EntityMapping Methods:
+ `boolean isSuperClass(Mapping context, ClassOutline theClassOutline)`
+ `boolean containsNonDiscriminatorColumn(Entity entity)`

The second condition is required for an [EclipseLink][5] issue [#1918][95] that a SQL `INSERT` for tables with `IDENTITY` primary key must have at least one other column value. Often, this other column is the `DTYPE` column, alone. For example, a root entity that serves as a *one-to-many* container, needs the DTYPE column to avoid [EclipseLink-6023][96].

See sample project *Publication*: [README][3]; [zip][2].



[1]: https://github.com/patrodyne/hisrc-hyperjaxb/releases/tag/2.1.1
[2]: https://github.com/patrodyne/hisrc-hyperjaxb/releases/download/2.1.1/hisrc-hyperjaxb-ejb-sample-publication-2.1.1-mvn-src.zip
[3]: https://github.com/patrodyne/hisrc-hyperjaxb/blob/master/ejb/assembly/samples/publication/README.md
[5]: https://www.eclipse.org/eclipselink/
[94]: https://github.com/patrodyne/hisrc-hyperjaxb/commit/1a269e2895ee13049c92b38722f42aa80cf9ce17?diff=unified#diff-87ea0290cc2012ff242833e23e08ace1e5fd39e66d2652b2ebe112d42116b81b
[95]: https://github.com/eclipse-ee4j/eclipselink/issues/1918
[96]: https://javadoc.io/doc/org.eclipse.persistence/eclipselink/latest/constant-values.html#org.eclipse.persistence.exceptions.QueryException.OBJECT_TO_INSERT_IS_EMPTY

**Comment by patrodyne** [comment]
Fixed in [Release 2.1.1](https://github.com/patrodyne/hisrc-hyperjaxb/releases/tag/2.1.1).
